### PR TITLE
Introduce revocable user memory concept

### PR DESCRIPTION
**Pull request by pnowojski**
This is another iteration of https://github.com/prestodb/presto/pull/6280

This PR introduces concecpt of revocable memory. If operator reserves revocable memory it should be able to revoke this memory upon a request from presto.

As an example this PR includes usage of this concept in spill to disk in `HashAggregationOperator`.

CC @losipiuk

### CODE REVIEW

**Comment by dain** [comment]
* [ ] https://github.com/prestodb/presto/pull/7895/commits/d6d96ef3fb Improve exception messages
* [ ] https://github.com/prestodb/presto/pull/7895/commits/cf9052ca10 Refactor OperatorContext::reserveMemory
* [ ] https://github.com/prestodb/presto/pull/7895/commits/3b5968081f Make memory pool aware of revocable allocation type
* [ ] https://github.com/prestodb/presto/pull/7895/commits/3a55bb6081 Revocable system memory in contexts
* [ ] https://github.com/prestodb/presto/pull/7895/commits/6907a00cbf Check for blocked on revocable memory in Driver
* [ ] https://github.com/prestodb/presto/pull/7895/commits/867f880a6f Add start/finishMemoryRevoke() to Operator
* [ ] https://github.com/prestodb/presto/pull/7895/commits/aaabba961b Add memoryRevokingRequestedFuture to OperatorContext
* [ ] https://github.com/prestodb/presto/pull/7895/commits/1e48dffbac Introduce QueryContextVisitor
* [ ] https://github.com/prestodb/presto/pull/7895/commits/fbd5b09d3f Introduce MemoryRevokingScheduler
* [ ] https://github.com/prestodb/presto/pull/7895/commits/21c5729c1f Fix synchronization of OperatorContext revocable memory state
* [ ] https://github.com/prestodb/presto/pull/7895/commits/20bd8af01c Requesting revoking unblocks driver
* [ ] https://github.com/prestodb/presto/pull/7895/commits/bdbcab1e2f Revoke memory in Driver
* [ ] https://github.com/prestodb/presto/pull/7895/commits/dbbac2fac6 Cleanup TestMemoryPools test
* [ ] https://github.com/prestodb/presto/pull/7895/commits/b374ea92c3 Add revoking tests to TestMemoryPools
* [ ] https://github.com/prestodb/presto/pull/7895/commits/50811ef4ca Trigger memory revoking in OperatorAssertions
* [ ] https://github.com/prestodb/presto/pull/7895/commits/23bb03a1c6 Add custom TaskContext builder
* [ ] https://github.com/prestodb/presto/pull/7895/commits/4980c5742f Always revoke memory in TestHashAggregationOperator
* [ ] https://github.com/prestodb/presto/pull/7895/commits/06c8096861 Always revoke memory in TestLocalBinarySpilledQueries
* [ ] https://github.com/prestodb/presto/pull/7895/commits/7d99d01854 Configure spill space limits in LocalQueryRunner
* [ ] https://github.com/prestodb/presto/pull/7895/commits/6853954c53 Always revoke memory in TestQuerySpillLimits
* [ ] https://github.com/prestodb/presto/pull/7895/commits/150c7cf57e Use revocable memory for spilling in aggregation
* [ ] https://github.com/prestodb/presto/pull/7895/commits/b2327d2a68 Rename memoryLimitBeforeSpill in HashAggregationOperator
* [ ] https://github.com/prestodb/presto/pull/7895/commits/11edb0b22a Prevent multiple calls to DriverContext.failed
* [ ] https://github.com/prestodb/presto/pull/7895/commits/212191b3fb Configurable memory revoking target and threshold
* [ ] https://github.com/prestodb/presto/pull/7895/commits/8f4316b780 Extract TaskManagementExecutor from SqlTaskManager
* [ ] https://github.com/prestodb/presto/pull/7895/commits/dbc2954a00 Trigger memory revoking quickly on over-allocation

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java [hunk 1]
```diff
@@ -171,9 +170,14 @@ public double getExpectedHashCollisions()
     }
 
     @Override
-    public ListenableFuture<?> isBlocked()
+    public ListenableFuture<?> startMemoryRevoke()
+    {
+        return null;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: shouldn't it rather throw? If we're here it means someone wants some memory back and we're giving none. It's an error.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 2]
```diff
@@ -160,9 +149,18 @@ private boolean hasPreviousSpillCompletedSuccessfully()
         }
     }
 
-    private boolean shouldSpill(long memorySize)
+    @Override
+    public ListenableFuture<?> startMemoryRevoke()
     {
-        return (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: Should the `memoryLimitBeforeSpill` variable in `com.facebook.presto.sql.planner.LocalExecutionPlanner.Visitor#visitAggregation` be removed too? I understand now there is no hard memory limit before spill.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java (hunk 1 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java]: I will throw `UnsupportedOperationException` from `startMemoryRevoke` and `finishMemoryRevoke`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java (hunk 2 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: I cleaned up the renaming commit.

**Comment by losipiuk** [comment]
@dain I addressed your initial commit and restructured the PR so it is easier for review.
I updated your check-list message.

File: presto-main/src/main/java/com/facebook/presto/operator/Operator.java [hunk 3]
```diff
@@ -60,6 +60,30 @@
...
+     * Spawned threads can not modify OperatorContext because it's not thread safe.
+     * For this purpose use implement finishMemoryRevoke
+     * <p>
+     * After startMemoryRevoke is called on Operator the Driver is disallowed to call any
+     * processing methods on it (finish/isFinished/isBlocked/needsInput/addInput/getOutput) until
+     * finishMemoryRevoke is called.
+     */
+    default ListenableFuture<?> startMemoryRevoke()
+    {
+        return NOT_BLOCKED;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Operator.java]: This is wrong default implementation. This method is called when `Operator` _has_ some revocable memory and should release it. Default implementation should be:
```
throw new UnsupportedOperationException(format("%s allocates revocable memory but does not implement revoking", getClass()));
```

**Comment by dain** [comment]
* [ ] https://github.com/prestodb/presto/pull/7895/commits/0edc644d59 Improve exception messages
* [ ] https://github.com/prestodb/presto/pull/7895/commits/2de4a3500b Refactor OperatorContext::reserveMemory
* [ ] https://github.com/prestodb/presto/pull/7895/commits/0db149803d Make memory pool aware of revocable allocation type
* [ ] https://github.com/prestodb/presto/pull/7895/commits/1175d0ec9d Revocable system memory in contexts
* [ ] https://github.com/prestodb/presto/pull/7895/commits/316548e75c Fix race-condition in Driver checking memory overuse
* [ ] https://github.com/prestodb/presto/pull/7895/commits/f328f92e08 Handle operator memory revoking in Driver
* [ ] https://github.com/prestodb/presto/pull/7895/commits/1ed2898003 Introduce QueryContextVisitor
* [ ] https://github.com/prestodb/presto/pull/7895/commits/b6bd52f682 Extract TaskManagementExecutor from SqlTaskManager
* [ ] https://github.com/prestodb/presto/pull/7895/commits/7830504113 Introduce MemoryRevokingScheduler
* [ ] https://github.com/prestodb/presto/pull/7895/commits/c906c0ad09 fixup! Introduce MemoryRevokingScheduler
* [ ] https://github.com/prestodb/presto/pull/7895/commits/98058c5704 Trigger memory revoking quickly on over-allocation
* [ ] https://github.com/prestodb/presto/pull/7895/commits/1d1183244e Cleanup TestMemoryPools test
* [ ] https://github.com/prestodb/presto/pull/7895/commits/c97b02ad79 Add revoking tests to TestMemoryPools
* [ ] https://github.com/prestodb/presto/pull/7895/commits/7fdd82d6c3 Trigger memory revoking in OperatorAssertions
* [ ] https://github.com/prestodb/presto/pull/7895/commits/b527557873 Add custom TaskContext builder
* [ ] https://github.com/prestodb/presto/pull/7895/commits/c55ed88a20 Always revoke memory in TestHashAggregationOperator
* [ ] https://github.com/prestodb/presto/pull/7895/commits/874b7344a8 Always revoke memory in TestLocalBinarySpilledQueries
* [ ] https://github.com/prestodb/presto/pull/7895/commits/6a0a95f232 Configure spill space limits in LocalQueryRunner
* [ ] https://github.com/prestodb/presto/pull/7895/commits/38a9ebbfe9 Always revoke memory in TestQuerySpillLimits
* [ ] https://github.com/prestodb/presto/pull/7895/commits/31c2f0a6f1 Use revocable memory for spilling in aggregation
* [ ] https://github.com/prestodb/presto/pull/7895/commits/01f524c671 Rename memoryLimitBeforeSpill config/session property
* [ ] https://github.com/prestodb/presto/pull/7895/commits/c49abf37b9 Update memory allocation when finishing in HashAggregationOperator
* [ ] https://github.com/prestodb/presto/pull/7895/commits/5e7bd860c1 Do not print configuration in tests to avoid Travis log limit
* [ ] https://github.com/prestodb/presto/pull/7895/commits/bc28da20d3 Add TestDistributedSpilledQueries

**Comment by losipiuk** [comment]
@dain One request. Could you please use GitHub's `Review changes` mechanism; instead commenting on individual commits. Adding single-comments on commits generates separate email notification thread for each commit from PR and it is hard to follow.
Thanks.

**Comment by dain** [comment]
@losipiuk I'm sorry.  I noticed too late that I somehow ended up commenting on disconnected commits.  I'll finish the review using the tool.

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        assertEquals(1, memoryPool.getFreeBytes());
+        scheduler.requestMemoryRevokingIfNeeded(tasks);
+        // we are still good - no revoking needed
+        assertMemoryRevokingNotRequested();
+
+        operatorContext4.setRevocableMemoryReservation(7);
+        assertEquals(-6, memoryPool.getFreeBytes());
+        scheduler.requestMemoryRevokingIfNeeded(tasks);
+        // we need to revoke 3 and 6
+        assertMemoryRevokingRequestedFor(operatorContext1, operatorContext3);
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: How do we guarantee the ordering this of revocation test?

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 5]
```diff
@@ -86,33 +103,125 @@ public MemoryRevokingScheduler(
...
     }
 
-    public void requestMemoryRevokingIfNeeded(Collection<SqlTask> sqlTasks)
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool)
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: As I mentioned in the other commit, I would make all of these methods synchronized.  This management code is fast enough that there should be no reason for it to be concurrent, which will make the job of future readers and maintainers simpler.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 6]
```diff
@@ -86,33 +103,125 @@ public MemoryRevokingScheduler(
...
+
+    @VisibleForTesting
+    void requestMemoryRevokingIfNeeded()
+    {
+        List<MemoryPool> poolsToCheck = memoryPools.stream()
+                .filter(this::memoryRevokingNeeded)
+                .collect(toList());
+        if (potentiallyOverflowingPools.addAll(poolsToCheck)) {
+            log.debug("Scheduling check for pools: ", poolsToCheck);
+            scheduleRevoking();
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: This is executed from a thread pool, so why wouldn't you just run this inline?

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 7]
```diff
@@ -86,33 +103,125 @@ public MemoryRevokingScheduler(
...
     {
-        memoryPools.forEach(memoryPool -> requestMemoryRevokingIfNeeded(sqlTasks, memoryPool));
+        try {
+            if (!memoryRevokingNeeded(memoryPool)) {
+                return;
+            }
+
+            if (potentiallyOverflowingPools.add(memoryPool)) {
+                log.debug("Scheduling check for %s", memoryPool);
+                scheduleRevoking();
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Is the check so slow that we can't do it on this thread?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java [hunk 8]
```diff
@@ -70,56 +74,93 @@ public synchronized MemoryPoolInfo getInfo()
...
+            }
+            else {
+                result = NOT_BLOCKED;
             }
-            checkState(!future.isDone(), "future is already completed");
-            return future;
         }
-        return NOT_BLOCKED;
+
+        onMemoryReserved();
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java]: Instead of notifying for every reservation (possibly a huge number), why not only notify when pool is "full", and if necessary when it becomes "not_full"?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.function.Consumer;
+
+public interface MemoryPoolListener
+{
+    /**
+     * Invoked when memory reservation completes successfully.
+     *
+     * @param memoryPool the {@link MemoryPool} where the reservation took place
+     */
+    default void onMemoryReserved(MemoryPool memoryPool) {}
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java]: No need for a default method.  If someone doesn't implement this, it is a pretty useless listener.

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Creates {@link MemoryPoolListener} implementing {@link #onMemoryReserved(MemoryPool)} only.
+     */
+    static MemoryPoolListener onMemoryReserved(Consumer<? super MemoryPool> action)
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java]: Why not just replace this interface with 'Consumer<? super MemoryPool>`?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java (hunk 8 above)
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java]: Add a memory pool test for this new behavior.

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 11]
```diff
@@ -221,6 +248,20 @@ private OperatorContext createContexts(SqlTask sqlTask)
...
     }
 
+    private void requestMemoryRevoking(MemoryRevokingScheduler scheduler)
+            throws Exception
+    {
+        scheduler.requestMemoryRevokingIfNeeded();
+        awaitAsynchronousCallbacksRun();
+    }
+
+    private void awaitAsynchronousCallbacksRun()
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: This is really hacky.  Can you add a counter to the `MemoryRevokingScheduler` and just verify the counter was incremented?

File: presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java [hunk 12]
```diff
@@ -65,15 +83,41 @@ public void setUp()
...
-        drivers = localQueryRunner.createDrivers("SELECT COUNT(*) FROM orders JOIN lineitem USING (orderkey)", outputFactory, taskContext);
+        setUp(() -> {
+            OutputFactory outputFactory = new PageConsumerOutputFactory(types -> (page -> { }));
+            return localQueryRunner.createDrivers("SELECT COUNT(*) FROM orders JOIN lineitem USING (orderkey)", outputFactory, taskContext);
+        });
+    }
+
+    private RevocableMemoryOperator setupConsumeRevocableMemory(DataSize reservedPerPage, long numberOfPages)
+    {
+        final AtomicReference<RevocableMemoryOperator> createOperator = new AtomicReference<>();
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java]: `final` shouldn't be necessary

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java [hunk 13]
```diff
@@ -56,4 +53,73 @@ public static TaskContext createTaskContext(QueryContext queryContext, Executor
...
+
+        private Builder(Executor executor, Session session)
+        {
+            this.executor = executor;
+            this.session = session;
+        }
+
+        public Builder setQueryMaxMemory(DataSize queryMaxMemory)
+        {
+            this.queryMaxMemory = queryMaxMemory;
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java]: check not null for all of these

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java [hunk 14]
```diff
@@ -186,8 +185,8 @@ public void testHashAggregation(boolean hashEnabled, long memoryLimitBeforeSpill
         assertOperatorEqualsIgnoreOrder(operatorFactory, driverContext, input, expected, hashEnabled, Optional.of(hashChannels.size()));
     }
 
-    @Test(dataProvider = "hashEnabled")
-    public void testHashAggregationWithGlobals(boolean hashEnabled)
+    @Test(dataProvider = "hashEnabledAndMemoryLimitBeforeSpillValues")
+    public void testHashAggregationWithGlobals(boolean hashEnabled, long memoryLimitBeforeSpill, long memoryLimitForMergeWithMemory)
```
Comment (dain) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java]: Maybe I'm missing it, but do we validate that the spill actually happened?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java [hunk 15]
```diff
@@ -63,8 +71,9 @@ public void testMaxSpillPerNodeLimit()
                 .put("experimental.max-spill-per-node", "10B")
                 .put("experimental.spiller-max-used-space-threshold", "1.0")
                 .build();
-        try (QueryRunner queryRunner = createDistributedQueryRunner(SESSION, properties)) {
-            queryRunner.execute(SESSION, "SELECT COUNT(DISTINCT clerk) as count, orderdate FROM orders GROUP BY orderdate ORDER BY count, orderdate");
+        try (QueryRunner queryRunner = createLocalQueryRunner(
+                new NodeSpillConfig().setMaxSpillPerNode(DataSize.succinctBytes(10)))) {
+            queryRunner.execute(queryRunner.getDefaultSession(), "SELECT COUNT(DISTINCT clerk) as count, orderdate FROM orders GROUP BY orderdate ORDER BY count, orderdate");
```
Comment (dain) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java]: Do we validate spill actually happens?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java [hunk 16]
```diff
@@ -63,8 +71,9 @@ public void testMaxSpillPerNodeLimit()
                 .put("experimental.max-spill-per-node", "10B")
                 .put("experimental.spiller-max-used-space-threshold", "1.0")
                 .build();
-        try (QueryRunner queryRunner = createDistributedQueryRunner(SESSION, properties)) {
-            queryRunner.execute(SESSION, "SELECT COUNT(DISTINCT clerk) as count, orderdate FROM orders GROUP BY orderdate ORDER BY count, orderdate");
+        try (QueryRunner queryRunner = createLocalQueryRunner(
```
Comment (dain) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java]: no need to wrap lines

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java [hunk 17]
```diff
@@ -79,24 +88,29 @@ public void testQueryMaxSpillPerNodeLimit()
                 .put("experimental.query-max-spill-per-node", "10B")
                 .put("experimental.spiller-max-used-space-threshold", "1.0")
                 .build();
-        try (QueryRunner queryRunner = createDistributedQueryRunner(SESSION, properties)) {
-            queryRunner.execute(SESSION, "SELECT COUNT(DISTINCT clerk) as count, orderdate FROM orders GROUP BY orderdate ORDER BY count, orderdate");
+        try (QueryRunner queryRunner = createLocalQueryRunner(
```
Comment (dain) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java]: no need to wrap

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class MemoryRevokingScheduler
+{
+    private static final Logger log = Logger.get(MemoryRevokingScheduler.class);
+
+    private static final Ordering<SqlTask> ORDER_BY_CREATE_TIME = Ordering.natural().onResultOf(task -> task.getTaskInfo().getStats().getCreateTime());
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Hmm. It should not starve as there should be some progress always. But I agree they could be slowed down considerably in case they are spilling constantly.
Generally ordering here was introduced mostly to make all of the nodes spill similar set of queries. So the global (cluster wide) set of queries that are spilling memory at the moment (hence blocked) is smaller. Using reversed ordering (newest first) seems similarly good for that purpose. But honestly I am not sure what are the runtime characteristics of this code without testing it in production under considerable load. And I am pretty sure that mechanism will need to be extended over time. 
E.g. we are thinking that when spilling starts for a query on one node, it is reasonable to schedule spilling of that query on other nodes ahead of time. 
Also it may be reasonable to take into consideration how much revocable memory is used by query when selecting the ones to spill. So we schedule spilling for fewer but heavier rather than more but lighter queries.

We can change that but I do not believe we have data right now to pick a solution which will surly be better. Any suggestions?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java (hunk 10 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPoolListener.java]: Listener interface is extensible, consumer is not. It's natural that you can add and remove listeners, which is the case here. I can change though, if you wish.

File: presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java (hunk 13 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingTaskContext.java]: This will be checked implicitly in builder's `build()`.

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java (hunk 14 above)
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashAggregationOperator.java]: you're right. we don't test that and now it's broken. I added temporary TODO note to unblock review process, to be done _before_ PR gets merged.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java (hunk 15 above)
Comment (findepi) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestQuerySpillLimits.java]: yes, with `@Test(expectedExceptions ...)`

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+                    }
+
+                    @Override
+                    public Void visitOperatorContext(OperatorContext operatorContext, AtomicLong remainingBytesToRevoke)
+                    {
+                        if (remainingBytesToRevoke.get() > 0) {
+                            long revokedBytes = operatorContext.requestMemoryRevoking();
+                            if (revokedBytes > 0) {
+                                remainingBytesToRevoke.addAndGet(-revokedBytes);
+                                log.info("(%s)requested revoking %s; remaining %s", memoryPool.getId(), revokedBytes, remainingBytesToRevoke.get());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: @losipiuk will create one :) What kind of stats would you see? How often (count revocation cases), amount?

**Comment by findepi** [comment]
@dain , comments applied. All as fixups but not all at the end. And there's one TODO we still need to do, but you can re-review already.

File: presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java [hunk 20]
```diff
@@ -113,71 +113,70 @@ public void tearDown()
...
-        DriverContext driverContext111 = pipelineContext11.addDriverContext();
-        OperatorContext operatorContext1 = driverContext111.addOperatorContext(1, new PlanNodeId("na"), "na");
-        OperatorContext operatorContext2 = driverContext111.addOperatorContext(2, new PlanNodeId("na"), "na");
-        DriverContext driverContext112 = pipelineContext11.addDriverContext();
-        OperatorContext operatorContext3 = driverContext112.addOperatorContext(3, new PlanNodeId("na"), "na");
+        TaskContext q1 = sqlTask1.getQueryContext().addTaskContext(new TaskStateMachine(new TaskId("q1", 1, 1), executor), session, false, false);
+        PipelineContext q1_p1 = q1.addPipelineContext(0, false, false);
+        DriverContext q1_p1_d1 = q1_p1.addDriverContext();
+        OperatorContext q1_p1_d1_o1 = q1_p1_d1.addOperatorContext(1, new PlanNodeId("na"), "na");
+        OperatorContext q1_p1_d1_o2 = q1_p1_d1.addOperatorContext(2, new PlanNodeId("na"), "na");
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/execution/TestMemoryRevokingScheduler.java]: I will rollback this change because those variable names are rejected by checkstyle.

File: presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java (hunk 19 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java]: Added #8313

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java (hunk 8 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryPool.java]: Done

**Comment by findepi** [comment]
@dain , comments applied & rebased on current master. All new code is as fixup commits, but not necessarily at the end (to avoid problems when merging)

**Comment by losipiuk** [comment]
@dain. Do you still want to take a look a this one? Or can we move forward? This is not very invasive. And if spilling is not explicitly enabled in config the system behaviour should be intact.

**Comment by dain** [comment]
I read through my comments and the replies, and am satisfied.  @losipiuk you can move forward without me

**Comment by ArturGajowy** [comment]
🎆
